labscript/labscript.py

```diff
@@ -2198,18 +2286,19 @@ def save_labscripts(hdf5_file):
                         # Doesn't seem to want to double count files if you just import the contents of a file within a module
                         continue
                     hdf5_file.create_dataset(save_path, data=open(path).read())
-                    if compiler.save_hg_info:
-                        hg_commands = [['log', '--limit', '1'], ['status'], ['diff']]
-                        for command in hg_commands:
-                            process = subprocess.Popen(['hg'] + command + [os.path.split(path)[1]], cwd=os.path.split(path)[0],
-                                                       stdout=subprocess.PIPE, stderr=subprocess.PIPE, startupinfo=startupinfo)
-                            info, err = process.communicate()
-                            if info or err:
-                                hdf5_file[save_path].attrs['hg ' + str(command[0])] = info.decode('utf-8') + '\n' + err.decode('utf-8')
+                    with _vcs_cache_rlock:
+                        if path not in _vcs_cache:
```

Comment:
Similarly here, you could do:

```
with _vcs_cache_rlock:
    condition = path not in _vcs_cache
if condition:
    #add to filewatch, and call callback
with _vcs_cache_rlock:
    for command, info, err in _vcs_cache[path]:
```

to reduce lock contention. This actually might mean I was wrong about needing the `RLock` over a plain `Lock` (but whatever...`RLock`s are safer in the long run).

Reply:
When I did it this way I thought I was just making things more complicated than necessary. As soon as the lock is released after the first `with` statement, it is almost immediately re-acquired either by the explicit callback call or the following `with` statement.

But now that you brought it up again, I think you're right that this is better. Freeing up the lock even briefly gives the `_file_watcher` thread an opportunity to run `_file_watcher_callback()` between those statements so it could potentially update the cache last minute right before last `with` statement there. I'll make this change.

I also considered reducing lock contention even further by locking just to do `file_cache_entry = _vcs_cache[path].copy()` then freeing the lock during `for` loop which is slowed by writing to disk. I think that would work fine as long as nothing that in the list returned by `_vcs_cache[path]` is mutable and edited by both threads. Currently the list only has tuples of strings so that should be ok unless something changes. Alternatively `copy.deepcopy()` could be used to play it safe. Thoughts?

Reply:
Commit 4716ceb implements your suggestion, but doesn't move the `for` loop out of the locked section as mentioned above.